### PR TITLE
LinuxBoot Support in UefiPayload

### PR DESCRIPTION
**Pull request by chengchiehhuang**
This PR includes patches to support Linuxboot in UefiPayload.

Ref:
 1. [u-root command](https://github.com/u-root/u-root/tree/master/cmds/exp/uefiboot)
 2. [doc](https://docs.google.com/document/d/1mU6ICHTh0ot8U45uuRENKOGI8cVzizdyWHGYHpEguVg/edit#heading=h.xzptrog8pyxf)

### CODE REVIEW

File: UefiPayloadPkg/UefiPayloadEntry/UefiPayloadEntry.c [hunk 1]
```diff
@@ -517,6 +517,8 @@ BuildGenericHob (
 
   // The UEFI payload FV
   BuildMemoryAllocationHob (PcdGet32 (PcdPayloadFdMemBase), PcdGet32 (PcdPayloadFdMemSize), EfiBootServicesData);
+  // The UEFI payload config FV
+  BuildMemoryAllocationHob (PcdGet32 (PcdPayloadFdMemBase) - SIZE_64KB, SIZE_64KB, EfiRuntimeServicesData);
```
Comment (gdong1) [inline UefiPayloadPkg/UefiPayloadEntry/UefiPayloadEntry.c:521]: why build runtime memory hob here? For other information, the PayloadEntry should already got the information and build a new HOB list to DXE core. Even it is really needed, maybe you need add LINUXBOOT_PAYLOAD flag for this code.

**Comment by chengchiehhuang** [comment]
Hi gdong1,

I am not sure if you prefer using PR for review or git sendmail. I am kind of doing both right now :).

File: UefiPayloadPkg/UefiPayloadEntry/UefiPayloadEntry.c (hunk 1 above)
Comment (chengchiehhuang) [reply UefiPayloadPkg/UefiPayloadEntry/UefiPayloadEntry.c:522]: fix per mail discussion.

File: UefiPayloadPkg/Library/LbParseLib/LbParseLib.c [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    return config;
+
+  DEBUG((DEBUG_ERROR,
+         "Expect payload config version %016lx or %016lx, but get %016lx\n",
+         UEFI_PAYLOAD_CONFIG_VERSION1, UEFI_PAYLOAD_CONFIG_VERSION2,
+         config->Version));
+  CpuDeadLoop();
+  while (1)
+    ;
+}
```
[the remark this answers is not in the document]
Comment (chengchiehhuang) [reply UefiPayloadPkg/Library/LbParseLib/LbParseLib.c:32]: For Trammell CLs, we will send another CL to include his change. This PR will aim for minimal support.

File: UefiPayloadPkg/Library/LbParseLib/LbParseLib.c [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+#include <Library/BlParseLib.h>
+#include <Library/DebugLib.h>
+#include <Library/IoLib.h>
+#include <Library/PcdLib.h>
+#include <Linuxboot.h>
+#include <Uefi/UefiBaseType.h>
+#include <stdint.h>
+#include <stdlib.h>
+//#include <string.h>
+//#include <ctype.h>
```
[the remark this answers is not in the document]
Comment (chengchiehhuang) [reply UefiPayloadPkg/Library/LbParseLib/LbParseLib.c]: will submit this in another PR.

File: UefiPayloadPkg/Include/Linuxboot.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+// magic version config is "LnxBoot1"
+#define UEFI_PAYLOAD_CONFIG_VERSION1 1
+#define UEFI_PAYLOAD_CONFIG_VERSION2 0x31746f6f42786e4cULL
+
+#define LINUXBOOT_MEM_RAM 1
+#define LINUXBOOT_MEM_DEFAULT 2
+#define LINUXBOOT_MEM_ACPI 3
+#define LINUXBOOT_MEM_NVS 4
+#define LINUXBOOT_MEM_RESERVED 5
+
```
[the remark this answers is not in the document]
Comment (chengchiehhuang) [reply UefiPayloadPkg/Include/Linuxboot.h]: done

**Comment by chengchiehhuang** [comment]
personally, I think doing this in PR is much easier for me :)

**Pull request closed without merging**